### ISSUE TITLE
Undefined offset: 0 in .../vendor/phpseclib/phpseclib/phpseclib/Net/SSH2.php on line 2596

### ISSUE THREAD

**Issue by zerkms** [phpseclib/phpseclib#1174]
The library sporadically throws

    Undefined offset: 0 in .../vendor/phpseclib/phpseclib/phpseclib/Net/SSH2.php on line 2596

Other applications successfully connect with no issues ever, but phpseclib randomly (not frequently) fails with that notice.

**Comment by bantu** [comment]
Which version of phpseclib are you using?

**Comment by zerkms** [comment]
Sorry, it's the latest stable v2.0.6

And as I mentioned - it's intermittent and really tricky to reproduce.

**Comment by bantu** [comment]
So that would be https://github.com/phpseclib/phpseclib/blob/2.0.6/phpseclib/Net/SSH2.php#L2596

@terrafrost 
Maybe `isset($this->server_channels[self::CHANNEL_EXEC])` should be checked.
Looks like it would have to be filled by https://github.com/phpseclib/phpseclib/blob/2.0.6/phpseclib/Net/SSH2.php#L3418.

**Comment by terrafrost** [comment]
What would be really helpful are server logs for when it does happen. As is it's not immediately clear to me how this could even happen.

In `exec` there's this:

```php
        $this->channel_status[self::CHANNEL_EXEC] = NET_SSH2_MSG_CHANNEL_OPEN;

        $response = $this->_get_channel_packet(self::CHANNEL_EXEC);
```

And then later, in `_get_channel_packet`, this:

```php
                switch ($this->channel_status[$channel]) {
                    case NET_SSH2_MSG_CHANNEL_OPEN:
                        switch ($type) {
                            case NET_SSH2_MSG_CHANNEL_OPEN_CONFIRMATION:
...
                                $this->server_channels[$channel] = $server_channel;
...
                            default:
                                user_error('Unable to open channel');
                                return $this->_disconnect(NET_SSH2_DISCONNECT_BY_APPLICATION);
```
99% of the time when `_get_channel_packet` has an issue it returns false. The only two exceptions to this appear to be as follows:

```php
            if ($client_channel == -1 && $response === true) {
                return true;
            }
            if (!strlen($response)) {
                return '';
            }
```
The first one wouldn't apply since self::CHANNEL_EXEC isn't -1. As for the `!strlen($response)`...  any packet that might have a zero length response should be filtered away in `_filter`, which is called at the end of `_get_binary_packet`.

Anyway, as my analysis shows, it's not clear how this is happening, so the server logs would help.

> Maybe `isset($this->server_channels[self::CHANNEL_EXEC])` should be checked.

That feels like that would just be sweeping the problem under the rug and not really fixing it. Kinda like putting the error suppression on every function call or putting an entire class in a giant try / catch (Exception $e) block.

**Comment by terrafrost** [comment]
TLDR the server logs would help. They can be obtained by doing `define('NET_SSH2_LOGGING', 2);` at the top of the file and then `echo $ssh->getLog();` after you encounter the issue.

**Comment by zerkms** [comment]
Right, I will add the corresponding changes and hopefully adding logging does not turn this into the observer's paradox.
Thanks.

**Comment by Nilithus** [comment]
Just wanted to report to also seeing this issue in version 1.0.5 the other day.

`Undefined offset: 0 .../vendor/phpseclib/phpseclib/phpseclib/Net/SSH2.php:2577`

which would be [this line](https://github.com/phpseclib/phpseclib/blob/1.0.5/phpseclib/Net/SSH2.php#L2577) Unfortunately we did not have `NET_SSH_LOGGING` defined. So we weren't able to capture any logs.

If we find anything note worthy that explains it will let you know.

**Comment by royduin** [comment]
Same issue here, but `Undefined offset: 1` in vendor/phpseclib/phpseclib/phpseclib/Net/SSH2.php line 2766

**Comment by terrafrost** [comment]
@royduin - what version of phpseclib are you using? What's on line 2766 likely varies from version to version.

Thanks!

**Comment by royduin** [comment]
I'm currently using version 2.0.7

**Comment by terrafrost** [comment]
@royduin - can you give me an overview of what your script does? Like does your script do just one `$ssh->exec('...');` and that's it or are you uploading stuff via SFTP or what?

**Comment by royduin** [comment]
I run scripts to check my servers, these commands run pretty often (some every minute, others every 5 minutes on 40 servers) so one of them is probably causing this error:

- Primary disk usage percentage: `df -h | grep ^/dev/ | head -1 | awk '{print $5}' | tr -dc '0-9'`
- Secondary disk usage percentage: `if [ $(df -h | grep ^/dev/ | wc -l) -eq 2 ]; then df -h | grep ^/dev/ | tail -1 | awk '{print $5}' | tr -dc '0-9'; fi`
- Used memory percentage: `free | grep 'buffers/cache' | awk '{print 100 - $4/($3+$4) * 100}' | grep -Eo '[0-9]+\.[0-9]+' || free | grep 'Mem' | awk '{print $6/$2 * 100}') | awk '{print int($1)}'`
- Used CPU percentage: `top -d 0.5 -b -n2 | grep "Cpu(s)"|tail -n 1 | awk '{print $2 + $4}' | awk '{print int($1)}'`

But when I look in Sentry this error appeared 8 times 1 one month. When I look at the times they occurred it's always between 6:30 and 6:55 (Central European Time). At 6:30 AutoMySQLBackup is starting to backup all the databases on the servers so I think the servers are pretty busy and don't always return the expected response.

**Comment by terrafrost** [comment]
@royduin - I'd say enable logging and then apply this change:

```php
#
#-----[ OPEN ]------------------------------------------
#
Net/SSH2.php
#
#-----[ FIND ]------------------------------------------
# in exec()
#
        $packet = pack(
            'CNNa*CNa*',
            NET_SSH2_MSG_CHANNEL_REQUEST,
            $this->server_channels[self::CHANNEL_EXEC],
            strlen('exec'),
            'exec',
            1,
            strlen($command),
            $command
        );
#
#-----[ BEFORE, ADD ]-----------------------------------
#
    file_put_contents('log-' . date('Ymd') . '-' . uniqid() . '.log', $this->getLog());
```

**Comment by terrafrost** [comment]
Once you have that send me the log files as you get them.

Thanks!

**Comment by royduin** [comment]
I've implemented the logging a little bit different:
```
try {
    $value = $ssh->exec($command);
} catch (\Exception $e) {
    throw new \Exception($ssh->getLog());
}
```
Let's wait until it occurs again.

**Comment by terrafrost** [comment]
I'm not sure that's gonna work since that's not an exception you're getting - it's an error. Per http://php.net/manual/en/language.errors.php7.php PHP did change how errors work in PHP 7 but in my own testing you still can't catch stuff like that. Example:

```php
<?php
error_reporting(E_ALL);

try {
    echo $a[4];
} catch (Exception $e) {
    echo "caught\n";
}

try {
    echo $a[4];
} catch (Throwable $e) {
    echo "caught\n";
}

echo "all done!\n";
```

Neither Exception nor Throwable result in the error being caught.

Unless you're using something like [set_error_handler](http://php.net/manual/en/function.set-error-handler.php) I'm not sure it's gonna work the way you're wanting..
